pageserver/src/tenant.rs

```diff
@@ -3229,6 +3230,30 @@ impl Tenant {
     pub fn cached_synthetic_size(&self) -> u64 {
         self.cached_synthetic_tenant_size.load(Ordering::Relaxed)
     }
+
+    /// Flush any in-progress layers, schedule uploads, and wait for uploads to complete.
+    ///
+    /// This function can take a long time: callers should wrap it in a timeout if calling
+    /// from an external API handler.
+    pub async fn flush_remote(&self) -> anyhow::Result<()> {
+        let timelines = self.timelines.lock().unwrap().clone();
+
+        for (timeline_id, timeline) in timelines {
+            tracing::info!(%timeline_id, "Flushing...");
+            timeline.freeze_and_flush().await?;
+            tracing::info!(%timeline_id, "Waiting for uploads...");
+            if let Some(client) = &timeline.remote_client {
+                client.wait_completion().await?;
+            }
+        }
```

Comment:
This should probably be a joinset? I am thinking very close to this must've already existed but unsure why would you not use it here. I need to look at this more.

Reply:
Yep, you're right.  In a1b0e2b0f2fb72269f06f7679a18cec021a4e48a I've made this a JoinSet of tasks that hold Timeline::gate -- that provides a nice clean cancellation story by ensuring that Timeline::shutdown will wait for these to die (and they should die promptly as the enclosed I/O respects Timeline::cancel)